### PR TITLE
Transit Autounseal

### PR DESCRIPTION
**Pull request by chrishoffman**
This adds the ability to use a different Vault instance as an autounseal using Vault's Transit secret engine.

### CODE REVIEW

File: vault/seal/transit/transit.go [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+		var err error
+		disableRenewal, err = strconv.ParseBool(disableRenewalRaw)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var namespace string
+	switch {
+	case os.Getenv("VAULT_NAMESPACE") != "":
```
Comment (briankassouf) [inline vault/seal/transit/transit.go]: I think this should be `VAULT_TRANSIT_SEAL_NAMESPACE` to distinguish from the standard vault namespace var. Which potentially could be set on the server if the operator is using a vault client there.

File: vault/seal/transit/transit.go [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+		}
+		if client.Token() == "" {
+			return nil, errors.New("missing token")
+		}
+		s.client = client
+
+		if !disableRenewal {
+			// Renew the token immediately to get a secret to pass to renewer
+			secret, err := client.Auth().Token().RenewTokenAsSelf(s.client.Token(), 0)
+			// If we don't get an error renewing, set up a renewer.  The token may not be renewable or not have
```
Comment (briankassouf) [inline vault/seal/transit/transit.go]: Can you wrap this comment?

File: vault/seal/transit/transit.go [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+				go func() {
+					for {
+						select {
+						case err := <-renewer.DoneCh():
+							s.logger.Info("shutting down token renewal")
+							if err != nil {
+								s.logger.Error("error renewing token", "error", err)
+							}
+							return
+						case <-renewer.RenewCh():
```
Comment (briankassouf) [inline vault/seal/transit/transit.go]: We might want to pass some kind of shutdown chan in here so we can cancel this go routine if we are shutting down

File: vault/seal/transit/transit.go [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+				}()
+				go s.renewer.Renew()
+			} else {
+				s.logger.Info("unable to renew token, disabling renewal", "err", err)
+			}
+		}
+	}
+
+	// Send a value to test the seal and to set the current key id
+	if _, err := s.Encrypt(context.Background(), []byte("a")); err != nil {
```
Comment (briankassouf) [inline vault/seal/transit/transit.go]: We should do this before kicking off the renewer go routine

File: vault/seal/transit/transit.go [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+	case os.Getenv("VAULT_NAMESPACE") != "":
+		namespace = os.Getenv("VAULT_NAMESPACE")
+	case config["namespace"] != "":
+		namespace = config["namespace"]
+	}
+
+	apiConfig := api.DefaultConfig()
+	if config["address"] != "" {
+		apiConfig.Address = config["address"]
+	}
```
Comment (briankassouf) [inline vault/seal/transit/transit.go]: I think we also need to allow configuration of client TLS certs, it's likely that mTLS will be required.

File: vault/seal/transit/transit_acc_test.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+	pool, err := dockertest.NewPool("")
+	if err != nil {
+		t.Fatalf("Failed to connect to docker: %s", err)
+	}
+
+	dockerOptions := &dockertest.RunOptions{
+		Repository: "vault",
+		Tag:        "latest",
+		Cmd: []string{"server", "-log-level=trace", "-dev", fmt.Sprintf("-dev-root-token-id=%s", testToken),
```
Comment (briankassouf) [inline vault/seal/transit/transit_acc_test.go]: Adding a TLS config could help test out the TLS client cert settings

File: vault/seal/transit/transit.go [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+				if err != nil {
+					return nil, err
+				}
+				s.renewer = renewer
+
+				go func() {
+					for {
+						select {
+						case err := <-renewer.DoneCh():
+							s.logger.Info("shutting down token renewal")
```
Comment (briankassouf) [inline vault/seal/transit/transit.go]: So i wonder what we should do if we lose the token, if sealwrap is enabled we'd stop being able to return any wrapped value?

File: vault/seal/transit/transit.go (hunk 1 above)
Comment (chrishoffman) [reply vault/seal/transit/transit.go]: I looked at this originally and decided that vault being a client for another vault is pretty a pretty limited use case and there was no reason to treat it differently.  Also, considering we are using the default API config, I think this just falls default config for environment variables like VAULT_TOKEN etc.  I think this isn't much different.

File: vault/seal/transit/transit.go (hunk 2 above)
Comment (chrishoffman) [reply vault/seal/transit/transit.go]: I am not sure what you mean here.

File: vault/seal/transit/transit.go (hunk 3 above)
Comment (chrishoffman) [reply vault/seal/transit/transit.go]: We are stopping the renewer in the Finalize method which will fire the DoneCh.  I _think_ this should be sufficient.

File: vault/seal/transit/transit.go (hunk 7 above)
Comment (chrishoffman) [reply vault/seal/transit/transit.go]: Not really sure.  This really isn't much different if someone changes the creds on any autoseal, it just stops working sorta.

File: vault/seal/transit/transit.go (hunk 2 above)
Comment (briankassouf) [reply vault/seal/transit/transit.go]: Wrap it at 80 char line width

File: vault/seal/transit/transit.go [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+	if len(splitKey) != 3 {
+		return nil, errors.New("invalid ciphertext returned")
+	}
+	keyID := splitKey[1]
+	s.currentKeyID.Store(keyID)
+
+	ret := &physical.EncryptedBlobInfo{
+		Ciphertext: []byte(ciphertext),
+		KeyInfo: &physical.SealKeyInfo{
+			KeyID: keyID,
```
Comment (briankassouf) [inline vault/seal/transit/transit.go]: Should we be storing the keyName as well as the key term here?

File: vault/seal/transit/transit.go [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+		},
+	}
+	return ret, nil
+}
+
+// Decrypt is used to decrypt the ciphertext
+func (s *Seal) Decrypt(_ context.Context, in *physical.EncryptedBlobInfo) ([]byte, error) {
+	path := path.Join(s.mountPath, "decrypt", s.keyName)
```
Comment (briankassouf) [inline vault/seal/transit/transit.go]: Should we use the keyName that is stored in the SealKeyInfo?

File: vault/seal/transit/transit.go (hunk 8 above)
Comment (chrishoffman) [reply vault/seal/transit/transit.go]: This very much follows the mold or the other cloud auto unseals, the key that you are using is fixed while the version can change.  The only unseal that doesn't follow this mold is the enterprise HSM integration, but versions are usually not a construct that HSMs support.

File: vault/seal/transit/transit.go [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+	}
+
+	sealInfo := make(map[string]string)
+	sealInfo["address"] = s.client.Address()
+	sealInfo["mount_path"] = s.mountPath
+	sealInfo["key_name"] = s.keyName
+	if namespace != "" {
+		sealInfo["namespace"] = namespace
+	}
+
```
Comment (briankassouf) [inline vault/seal/transit/transit.go]: I think we should query the API to get the latest key term here and set the `currentKeyID` value. Probably best to just encrypt some dummy data so the ACL doesn't also need key read access.

**Pull request merged**